### PR TITLE
feat: add deployment context to connector factory

### PR DESCRIPTION
**Pull request by jhaeyaert**
**Issue**

https://graviteecommunity.atlassian.net/browse/APIM-266

**Description**

Add a deployment context object to pass to connector factory in order to get access to useful component such as api, services, ...

---
**Gravitee.io Automatic Deployment**

🚀 A prerelease version of this package has been published on Gravitee's private artifactory, you can:
 - use it directly by updating your project with version: `2.0.0-apim-266-http-proxy-connectors-SNAPSHOT`
 - download it from Artifactory [here](https://odbxikk7vo-artifactory.services.clever-cloud.com/gravitee-snapshots/io/gravitee/gateway/gravitee-gateway-api/2.0.0-apim-266-http-proxy-connectors-SNAPSHOT/gravitee-gateway-api-2.0.0-apim-266-http-proxy-connectors-SNAPSHOT.zip)

### CODE REVIEW

File: src/main/java/io/gravitee/gateway/jupiter/api/context/DeploymentContext.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public interface DeploymentContext {
+    /**
+     * Returns the component corresponding to the specified <code>componentClass</code>.
+     *
+     * @param componentClass the {@link Class} of the expected component to retrieve.
+     * @param <T> the expected instance type.
+     *
```
Comment (ytvnr) [inline src/main/java/io/gravitee/gateway/jupiter/api/context/DeploymentContext.java:30]: maybe we can provide some examples of components to retrieve ?

Comment (jhaeyaert) [reply src/main/java/io/gravitee/gateway/jupiter/api/context/DeploymentContext.java:30]: As it is more an implementation consideration it should not imho.

File: src/main/java/io/gravitee/gateway/jupiter/api/connector/ConnectorFactory.java [hunk 2]
```diff
@@ -37,9 +38,12 @@
     Set<ConnectorMode> supportedModes();
 
     /**
-     * Allow creating new connector from the given string configuration
+     * Allows creating new connector from the given string configuration.
```
Comment (guillaumelamirand) [inline src/main/java/io/gravitee/gateway/jupiter/api/connector/ConnectorFactory.java:41]:
```suggestion
     * Allow creating new connector from the given string configuration.
```
Allow was right actually. It is imperative.

File: src/main/java/io/gravitee/gateway/jupiter/api/connector/entrypoint/async/EntrypointAsyncConnectorFactory.java [hunk 3]
```diff
@@ -38,18 +39,22 @@ default ApiType supportedApi() {
...
      */
-    default EntrypointAsyncConnector createConnector(final String configuration) {
-        return createConnector(null, configuration);
+    default T createConnector(final DeploymentContext deploymentContext, final String configuration) {
+        return createConnector(deploymentContext, null, configuration);
     }
 
     /**
-     * Allow creating new connector from the given string configuration
+     * Allows creating new connector from the given string configuration and QoS (Quality Of Service).
```
Comment (guillaumelamirand) [inline src/main/java/io/gravitee/gateway/jupiter/api/connector/entrypoint/async/EntrypointAsyncConnectorFactory.java:51]:
```suggestion
     * Allow creating new connector from the given string configuration and QoS (Quality Of Service).
```

File: src/main/java/io/gravitee/gateway/api/ExecutionContext.java [hunk 4]
```diff
@@ -48,6 +48,12 @@ public interface ExecutionContext {
     String ATTR_ENVIRONMENT = ATTR_PREFIX + "environment";
     String ATTR_FAILURE_ATTRIBUTE = ATTR_PREFIX + "failure";
 
+    /**
+     * Skip security chain isn't prefixed to avoid breaking code where this constant isn't used.
```
Comment (guillaumelamirand) [inline src/main/java/io/gravitee/gateway/api/ExecutionContext.java:52]: As we are building the v4, shouldn't we start from a fresh/clean implementation without so much care about the legacy??

File: src/main/java/io/gravitee/gateway/jupiter/api/context/DeploymentContext.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ */
+package io.gravitee.gateway.jupiter.api.context;
+
+import io.gravitee.el.TemplateEngine;
+
+/**
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public interface DeploymentContext {
```
Comment (guillaumelamirand) [inline src/main/java/io/gravitee/gateway/jupiter/api/context/DeploymentContext.java:24]: Could you add some details here regarding the usage of DeploymentContext in opposite to ExecutionContext?

**Review by guillaumelamirand — CHANGES_REQUESTED** [review]

File: src/main/java/io/gravitee/gateway/api/ExecutionContext.java (hunk 4 above)
Comment (jhaeyaert) [reply src/main/java/io/gravitee/gateway/api/ExecutionContext.java:52]: I have created this attribute specifically for being able to not use it anymore 😜
In fact, no constant were defined (just string here and there) and I needed to support moving from the old attribute to the new internal attribute (have a look at the `ExecutionContextAdapter`).

File: src/main/java/io/gravitee/gateway/jupiter/api/context/DeploymentContext.java (hunk 5 above)
Comment (jhaeyaert) [reply src/main/java/io/gravitee/gateway/jupiter/api/context/DeploymentContext.java:27]: Done!

File: src/main/java/io/gravitee/gateway/api/ExecutionContext.java (hunk 4 above)
Comment (guillaumelamirand) [reply src/main/java/io/gravitee/gateway/api/ExecutionContext.java:52]: Ok I got it perfect !

**Review by guillaumelamirand — APPROVED** [review]

**Pull request merged**